fix(core): support `InjectFlags` argument in `NodeInjector.get()`

vthinkxie opened the pull request:
close #31776

the InectFlags was defined in both getOrCreateInjectable and Injector, but was ignored in NodeInjector
this PR support getting parent token via injector.get in NodeInjector

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A


## What is the new behavior?


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

Commit:
fix(core): support `InjectFlags` argument in `NodeInjector.get()`

The `InjectFlags` argument was defined for both `getOrCreateInjectable()` and `Injector`, but was ignored in `NodeInjector`.
This PR adds support for getting parent token via `NodeInjector.get()`.

close #31776

AndrewKushnir reviewed (commented):
@vthinkxie thanks for creating this PR. A couple comments:

- This change should also contain a number of tests that would make sure the flags work as expected and are aligned with token injection performed via constructor arguments.
- The `NodeInjector` is a part of the Ivy logic and the ViewEngine version of it doesn't seem to have the `flags` support either ([see here](https://github.com/angular/angular/blob/master/packages/core/src/view/refs.ts#L369)). We'd need to consider adding the logic there too (for general feature parity).

We'll discuss this with the team more and I will followup here with some additional information.

Thank you.

AndrewKushnir commented:
Hi @vthinkxie, sorry for the delay in answering.

We've discussed this feature request with the team and we'd like to move forward with the implementation. A couple notes:

- Passing flags from `NodeInjector.get` -> `getOrCreateInjectable` makes sense.
- This change requires a number of tests that would make sure the flags work as expected and are aligned with token injection performed via constructor arguments. You can add more tests in the `packages/core/test/acceptance/di_spec.ts` file.
- Since ViewEngine is deprecated, the change can be isolated to Ivy code only (i.e. what you've already changed).

Please let me know if any additional information is required (on adding tests or anything else related to this change).

Thank you.

vthinkxie commented on packages/core/src/di/injector.ts:
```diff
@@ -71,7 +71,7 @@ export abstract class Injector {
    * @suppress {duplicate}
    */
-  abstract get(token: any, notFoundValue?: any): any;
+  abstract get(token: any, notFoundValue?: any, flags?: InjectFlags): any;
```
add flags here in order to make flags testable

AndrewKushnir replied:
I think it'd be great to avoid changing deprecated method signatures. Could you please share which use-case in tests this affects?

vthinkxie replied:
https://github.com/angular/angular/pull/41592/files#diff-ece22a8d8ae1405b0b70f8824055ccc962cd82206fcdc9932dfd754885e31be3R1977

AndrewKushnir replied:
@vthinkxie thanks for the link. I think it'd be better to update the test code to use  the `new InjectionToken('token')` construction instead, so that call doesn't use deprecated signature (I will also post a note on the change itself with an example).

vthinkxie commented:
Hi @AndrewKushnir 

I have added some tests in `packages/core/test/acceptance/di_spec.ts`, could you have a look? thanks

AndrewKushnir commented on packages/core/test/acceptance/di_spec.ts:
```diff
@@ -1919,6 +1919,100 @@ describe('di', () => {
         });
       });
+
+      onlyInIvy('Ivy support InjectFlags in Injector')
```
It'd be great to also add a test where multiple flags are used, for example:

```
this.injector.get(TOKEN, null, InjectFlags.Host | InjectFlags.Optional);
```

vthinkxie replied:
I just add more some multiple flags test cases

AndrewKushnir commented on packages/core/test/acceptance/di_spec.ts:
```diff
@@ -1919,6 +1919,128 @@ describe('di', () => {
+              selector: 'parent',
+              template: '<child></child>',
+              providers: [{
+                provide: 'token',
```
In order to avoid the mentioned issue with the deprecated signature, try using the `InjectiorToken` here instead, for example:
```
  const TOKEN = new InjectionToken('token');

  // ... in @Component:
  provide: TOKEN,

  // ... in test code:
  expect(childComponent.injector.get(TOKEN)).toBe('CHILD');
```

AndrewKushnir commented on packages/core/test/acceptance/di_spec.ts:
```diff
@@ -1919,6 +1919,128 @@ describe('di', () => {
       });
+
+      onlyInIvy('Ivy support InjectFlags in Injector')
+          .it('should support Optional flag in Injector', () => {
+            @Directive({selector: '[dirA]'})
+            class DirectiveA {
+              dirB = this.injector.get(DirectiveB, null, InjectFlags.Optional);
+              constructor(public injector: Injector) {}
+            }
+            @Component({template: '<div dirA></div>'})
+            class MyComp {
+              @ViewChild(DirectiveA) dirA!: DirectiveA;
+            }
+
+            TestBed.configureTestingModule({declarations: [DirectiveA, DirectiveB, MyComp]});
+            const fixture = TestBed.createComponent(MyComp);
+            fixture.detectChanges();
+
+            const dirA = fixture.componentInstance.dirA;
+            expect(dirA.dirB).toBeNull();
+          });
```
I think you can simplify this test to have just one component:

```suggestion
          .it('should support Optional flag in Injector', () => {
            const NON_EXISTING_PROVIDER = new InjectionToken('non-existing');
    
            @Component({template: '...'})
            class MyComp {
              provider = null;
              constructor(public injector: Injector) {
                this.provider = this.injector.get(NON_EXISTING_PROVIDER, null, InjectFlags.Optional);
              }
            }

            TestBed.configureTestingModule({declarations: [MyComp]});
            const fixture = TestBed.createComponent(MyComp);
            fixture.detectChanges();

            expect(fixture.componentInstance.provider).toBeNull();
          });
```

AndrewKushnir commented on packages/core/test/acceptance/di_spec.ts:
```diff
@@ -1919,6 +1919,128 @@ describe('di', () => {
       });
+
+      onlyInIvy('Ivy support InjectFlags in Injector')
+          .it('should support Optional flag in Injector', () => {
```
```suggestion
          .it('should support Optional flag in NodeInjector', () => {
```

AndrewKushnir commented on packages/core/test/acceptance/di_spec.ts:
```diff
@@ -1919,6 +1919,128 @@ describe('di', () => {
         });
       });
+
+      onlyInIvy('Ivy support InjectFlags in Injector')
```
I think you can create a new `describe` section and put all tests inside:
```suggestion
      onlyInIvy('Ivy supports `InjectFlags` in NodeInjector').describe('`InjectFlags` support in NodeInjector', () => {
        // ... other tests go here ...
        // it('...', () => { ... });
      });
```

AndrewKushnir commented on packages/core/test/acceptance/di_spec.ts:
```diff
@@ -1919,6 +1919,128 @@ describe('di', () => {
+          });
+
+      onlyInIvy('Ivy support InjectFlags in Injector')
+          .it('should support Host flag in Injector', () => {
```
```suggestion
          .it('should support Host flag in NodeInjector', () => {
```

AndrewKushnir commented on packages/core/test/acceptance/di_spec.ts:
```diff
@@ -1919,6 +1919,128 @@ describe('di', () => {
+              }]
+            })
+            class ChildComponent {
+              constructor(public injector: Injector, @SkipSelf() public parentInjector: Injector) {}
```
I'd like to propose to refactor the code (in this and other tests) to do the following:
- inject a token via constructor params (using the necessary decorators, such as `@SkipSelf`, `@Optional`, etc)
- inject the same token via `injector.get` call
- compare results in test code

Here is a quick example:
```suggestion
              constructor(injector: Injector, @Inject('token') @SkipSelf() public tokenViaConstructor: string) {
                this.tokenViaInjector = injector.get('token', null, InjectFlags.SkipSelf);
              }
              
              // ... later in the test ...
              expect(childComponent.tokenViaConstructor).toBe('CHILD');
              expect(childComponent.tokenViaConstructor).toBe(childComponent.tokenViaInjector);
```

I believe that'd make tests more readable and would actually verify that constructor-based and `injector.get`-based behavior is the same.

vthinkxie commented:
Hi @AndrewKushnir thanks a lot for your review
I have updated the code, can you have a look again?

vthinkxie commented:
Hi @AndrewKushnir
thanks for your review, I have updated the code, could you take a look?

vthinkxie commented:
> Hi @vthinkxie, thanks for addressing the feedback. I will have a look at the changes within the next couple days and let you know.
> 
> One thing that I noticed is that the `ci/circleci: components-repo-unit-tests` CI job is failing. I've looked at the failure and found the code in Components repo that relies on the `Injector.get` flags, [see here](https://github.com/angular/components/blob/master/src/material/datepicker/date-range-input-parts.ts#L105). My guess is that the mentioned code assumes that the 3rd argument is supported (when it's actually ignored) and the change that you made in this PR makes use of that argument, thus changing the result. We will look into this more and let you know.
> 
> Thank you.

Thanks @AndrewKushnir ,hope this PR will be in time for the v12 release

AndrewKushnir commented:
> hope this PR will be in time for the v12 release

Unfortunately this change would not be included into v12 since the feature freeze (for v12) happened a couple weeks ago. We also had an additional discussion about the changes that implement Ivy-only features and our current understanding is that we can land these changes closer to v13 (when main branches are open for v13 changes). I've added this PR to the `v13-candidates` milestone, so we'll look at it again while working on v13 changes. Meanwhile we can still take this PR to a ready state, so it's less work later (and less risks in case more time is needed to complete all the changes). Currently the next steps are:
- resolve the issue with the `ci/circleci: components-repo-unit-tests` CI job (we are looking into that)
- perform another round of review after the most recent changes

We'll provide more updates in this PR once we have them. Thank you.

AndrewKushnir commented:
> resolve the issue with the ci/circleci: components-repo-unit-tests CI job (we are looking into that)

@vthinkxie just a quick update: the problem with the `ci/circleci: components-repo-unit-tests`  CI job was resolved in https://github.com/angular/components/pull/22665.

AndrewKushnir commented:
Hi @vthinkxie, just wanted to followup and let you know that master branch is available for v13 changes now and we can resume the process for this PR. Please let me know if you are still interested in landing this change. Thank you.

Azbesciak commented:
Hello, when this can be merged? I think that the issue with injector also propagates to further components if you create your own components via `ViewContainer.createComponent()`, not to mention that you cannot access parent providers via injector.

vthinkxie commented:
Hi @AndrewKushnir 
I have rebased onto the master, could you have a look?

DmitryEfimenko commented:
This didn't make it into v12. Maybe v13? 🤞

petebacondarwin commented:
The consensus on the team is that this is actually a bug rather than a feature.

Now that View Engine will be out the door in 13.0.0, we could land this more easily.

@vthinkxie  Please could you modify the commit message to use `fix` rather than `feat` and I will update the labelling, so that we should actually be able to land it in more relaxed versions.

petebacondarwin commented:
@vthinkxie do you not have time to work on this PR anymore?

petebacondarwin reviewed (approved):
LGTM - How about you @AndrewKushnir ?

petebacondarwin commented:
Oh there is a typo in the commit message. Can we update it to:

```
fix(core): support `InjectFlags` argument in `NodeInjector.get()`

The `InjectFlags` argument was defined for both `getOrCreateInjectable()` and `Injector`, but was ignored in `NodeInjector`.
This PR adds support for getting parent token via `NodeInjector.get()`.

Closes #31776
```

vthinkxie commented:
> Oh there is a typo in the commit message. Can we update it to:
> 
> ```
> fix(core): support `InjectFlags` argument in `NodeInjector.get()`
> 
> The `InjectFlags` argument was defined for both `getOrCreateInjectable()` and `Injector`, but was ignored in `NodeInjector`.
> This PR adds support for getting parent token via `NodeInjector.get()`.
> 
> Closes #31776
> ```

done

AndrewKushnir reviewed (approved):
@vthinkxie thanks for working on this change 👍

petebacondarwin reviewed (approved):
Let's merge it!
